### PR TITLE
Use iterative refinement guided by B-spline for conic drawing, cleanup and fix computation of k.

### PR DESCRIPTION
**Pull request by elkins0**
The code for computing `k` was changed to work optimally. This way, other then NaN, `k` absolutely cannot assume values other than between zero and one inclusively. Only one square root is needed this way, also helping with numerical precision. Changing the sign of all the `c##` no longer affects the computation of `k`.

### CODE REVIEW

**Comment by gagern** [comment]
I like what you did in the computation of `k`. And cleaning the old code, too; I totally forgot doing that. I'm somewhat skeptic about the changed termination criteria up front.

Why change `eps` (10<sup>−14</sup>) to `CSNumber.eps` (10<sup>−10</sup>)? Is there some specific problem this does solve? OK, this is the third coordinate, so if the other entries are large, then a small value there will indicate a point of intersection far out, justifying the use of lines instead. But is there some reason why we can expect the other values to be far out? For double lines, the dual conic would be all zero, so all entries of the pole would be zero. Which means that for situations close to double lines, like extremely flat ellipses or hyperbolas, we might be seeing small values in all coordinates. Perhaps not a problem, as drawing these using lines would likely be all right, too. Was your chance motivated by a consideration like this?

Why did you drop the `isFinite`  checks? As we normalized the original matrix, its entries are bounded by 1, so the `c##` are bounded by 2, so the `k##` are bounded by 20 or so. But the coordinates of the `u` are determined by the pixel coordinates of the endpoints, which means that for really large canvases I can see no hard upper bound on these. OK, infinities can crop up in various other places, too, and we don't check for all of them. I'm not sure why I considered the coordinates here more important. Have you checked that we can simply rely on NaN propagation to resolve this for us?

Why did you drop the `area` check? I added that after I noticed that the code would recurse on degenerate conics even though a single straight line would do. I haven't checked whether the same portion of the conic may get drawn more than once, but if that were the case, we might have a problem for conics drawn with an alpha value other than 1.

I haven't tested your code, and won't be able to before Christmas. But testing would only complement understanding, not replace it, so I'd like to know why you did these changes.

**Comment by elkins0** [comment]
@gagern: I'm pleased to see that you've made use of the adjoint matrix `k##` numbers and used the polar relation to get the control point.

>Why change eps (10<sup>-14</sup>) to CSNumber.eps (10<sup>-10</sup>)? 

The value 10<sup>-14</sup> greatly amplifies numerical precision errors giving chaotic results. You only have two decimal digits of precision left when you add 1.0+10<sup>-14</sup>. With 1.0+10<sup>-10</sup> you get six. 

For example, agonize over this, the order in which three values are added affects which result you get because 1.0+10<sup>-16</sup> leaves you with no digits of precision at all, only a single bit (double precision floating point math plays tricks on you):
```
1.0-1.0+1.0e-16===1.0e-16
1.0+1.0e-16-1.0===0.0
Math.pow(2,-53)===1.1102230246251565e-16
Math.pow(2,-53)!==1.0e-16
(1.0-1.0+1.0e-16)/2+(1.0+1.0e-16-1.0)/2===5.0e-17
(1.0-1.0+2.0e-16)/2+(1.0+2.0e-16-1.0)/2===Math.pow(2,-53)+1.0e-16
```
Remember that you only get 53 bits of precision in the mantissa of a double precision floating point number which works out to a little less than 16 decimal digits of precision:
`Math.pow(2,53)===0.9007199254740992e16`

>Why did you drop the `isFinite` checks?

I don't believe there is there a reason to expect that `pt1`, `pt2` or `mat` should contain `NaN` values at this point in the code. Since we are avoiding the divide-by-zero of `cz` and `pt1` and `pt2` are on or inside the boundaries they should already be finite, certainly an expression involving them should give  nothing approaching infinity, so the `isFinite` checks are superfluous here.

>Why did you drop the `area` check?

The code, as you have it now, assures that the middle of the Bézier curve is within a distance of less than two tenths of a pixel of the actual curve. In this diagram you can see that distance the desired versus actual curves are apart varies widely for the same given area. In the first two cases the distance they are apart remains the same despite the area being reduced by one third. In the last case they are only one pixel apart and yet the area is still very large.
canvas
My question is, why did you choose to be within a distance of less than two tenths of a pixel of the actual curve instead of less than one half of a pixel? Do you have a case showing this to be detectable without using a magnifying glass?

**Comment by gagern** [comment]
Thanks for the detailed explanations!

> Why change `eps` (10<sup>−14</sup>) to `CSNumber.eps` (10<sup>−10</sup>)?

I can see how 10<sup>−10</sup> is essentially zero *if* added to (or subtracted from) another number of magnitude approximately one. But you're using this ε check for `cz`, which is the sole denominator of a division. I guess one *could* argue that if the value of `cz` is this small, then it must be the *result* of highly imprecise calculations leading to that. But is that really true? can't we have small entries of the primal matrix combining to even smaller entries of the dual, leading to a well-defined and stable point `c` despite the small denominator? On the other hand, one could say that if the denominator is this small, then the point `c` as a whole is either poorly defined or far out, and in either case drawing a straight line is probably the best thing to do. So I think I'm fine with this now.

> Why did you drop the `area` check?

You have already convinced me that the area *by itself* isn't as good an estimate for quality of approximation as I originally thought. However, I think that in the absence of this check we need to find a way to e.g. avoid the large number of quadratic segments used for scenarios like the following one:

    ["degen9", [0,1,0.25], [0,1,0.3333333333333333], [0,1,-0.5], [0.29746835443037983,0.9999999999999999,-0.1582278481012658], [1,0,-0.16666666666666666]],

degen9

This *should* be drawn using 4 line segments, but actually we get like 35 quadratic curves.

Originally I had feared that the algorithm might draw parts of the conic more than once, or might have even more serious problems if the point of intersection lies outside the canvas, but I could observe neither of these in my tests. So apart from the large number of primitive draw operations, I can see no immediate drawbacks from not having the area check.

> why did you choose to be within a distance of less than two tenths of a pixel

That's because I have HiDPI devices and exported files in mind. Our canvas uses CSS pixel coordinates, but a HiDPI device might well have up to three physical device pixels per logical CSS pixel. And given the way the industry evolves, I wouldn't be surpised to see device pixel ratios of 4 or even 5 in the next few years. So I decided on a limit of ⅕ pixel to not be more than one pixel off even on a theoretical DPR 5 device.

I also like to use the same code path for all devices, instead of making this limit here device-dependent. That avoids surprises due to insufficient testing with different devices. It also has the benefit that a vector graphics file (PDF or SVG) exported from CindyJS will work the same no matter the device pixel ratio of the browser used when exporting the image. If the image gets printed to paper, even small differences between the curve of a conic and the position of a point on that conic might become visible, I think. Even more so if the image gets enlarged, e.g. to be placed on a scientific poster.

**Comment by elkins0** [comment]
@gagern:
>However, I think that in the absence of this check we need to find a way to e.g. avoid the large number of quadratic segments...

I've implemented code changes 45f2e40d54698ae8f3a6b69e4e218bbca81d5c49 and b348b59bc46033c29236a1ad594117ed47816665 to take care of that. Basically, if the `G` point is near the midpoint `M`, a line segment is drawn from `A` to `C`, when `G` is near the control point `B`, two segments draw the steeple from `A` to `B` to `C`.

*Updated*
Fixed to use (G-M)^2 not (B-M)^2 by adding k² factor since in parametric terms `G==k*B+(1-k)*M` so `G-M==k*B-k*M==k*(B-M)` and `B-G==(1-k)*B-(1-k)*M==(1-k)*(B-M)`.

File: src/js/libcs/OpDrawing.js [hunk 1]
```diff
@@ -374,12 +374,21 @@ function solveRealQuadraticHomog(a, b, c) {
...
-// Returns either null (if solutions would be complex or NaN)
+// Returns either null (if both solutions would be complex or NaN)
 // or two values x satisfying ax^2 + bx + c = 0
 function solveRealQuadratic(a, b, c) {
-    var hom = solveRealQuadraticHomog(a, b, c);
-    if (hom === null) return null;
-    return [hom[0][0] / hom[0][1], hom[1][0] / hom[1][1]];
+    //var hom = solveRealQuadraticHomog(a, b, c);
+    //if (hom === null) return null;
+    //return [hom[0][0] / hom[0][1], hom[1][0] / hom[1][1]];
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js]: I think you can drop these comments. Out of curiosity, what was your motivation to inline the code here, instead of delegating the common functionality to the homogeneous version?

File: src/js/libcs/OpDrawing.js [hunk 2]
```diff
@@ -479,39 +509,23 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
     // Sign 1 means inside, i.e. polar has complex points of intersection.
     // Sign -1 means outside, i.e. polar has real points of intersection.
     // Sign 0 would be on conic, but numeric noise will drown those out.
     // Note that this distinction is arbitrary for degenerate conics.
     function sign(x, y) {
-        var s = (c20 * x + c11 * y + c10) * x + (c02 * y + c01) * y + c00;
-        if (s >= 0) return 1;
-        if (s < 0) return -1;
-        return NaN;
+        return (c20 * x + c11 * y + c10) * x + (c02 * y + c01) * y + c00;
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js:531]: If you return the result of the quadratic form, instead of just the sign, then you should change the name and description of the function.

File: src/js/libcs/OpDrawing.js [hunk 3]
```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
+            var extent1 = sort > 0 ? extent : 0;
+            // Only include each first corner that falls within the conic
+            if (bInside) {
+                t = "corner";
+                coord = extent0;
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+            coord = sort > 0 ? sol[0] : sol[1];
+            if (0 < coord && coord < extent || coord === extent1) {
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js]: Comparing the result of a floating point computation using exact equality sounds like asking for trouble. The problematic scenario here would be the following: a conic passes through the corner of the rectangle, but due to numeric errors, the intersections on both the horizontal and the vertical edge are both deemed outside the rectangle. In that case you don't notice that you crossed the conic at all. That's the kind of scenario I was trying to avoid by using the signs at the corners. I don't rely on whether the solutions are inside or outside the edge segment, but instead I rely on whether the solutions result in the expected sign changes or not.

File: src/js/libcs/OpDrawing.js [hunk 4]
```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
+                t = bInside ? "begin" : "end";
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+        } else {
+            var x, y;
+            var flat = flats[vert ? 0 : 1];
+            if (flat !== null) {
+                t = "split";
+                flat = flat[other === 0 ? 0 : 1];
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js]: Using *either* the tangency points *or* the intersection points is an interesting approach which avoids some case distinctions and the sorting issues I had to worry with. :+1:

File: src/js/libcs/OpDrawing.js [hunk 5]
```diff
@@ -303,7 +303,7 @@ eval_helper.drawcircle = function(args, modifs, df) {
         [-xx, -yy, xx * xx + yy * yy - rad * rad]
     ]);
 
-    eval_helper.drawconic(cMat, modifs);
+    eval_helper.drawconic(cMat, modifs, df);
     */
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js:307]: We might want to delete this whole comment, particularly seeing how the statement “drawconic doesn't handle filling” is no longer true. The comment dates from a5d0f6796095d8571d671e91ab208bc0c45016e3 but the commented-out code was superseded by #262 and later on by a fix for [Chrome issue #587582](https://bugs.chromium.org/p/chromium/issues/detail?id=587582).

File: src/js/libcs/OpDrawing.js [hunk 6]
```diff
@@ -355,7 +355,7 @@ evaluator.drawconic$1 = function(args, modifs) {
         }
 
     }
-    return eval_helper.drawconic(arr, modifs);
+    return eval_helper.drawconic(arr, modifs, "D");
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js:358]: A similar change is required [in GeoRenderer](https://github.com/CindyJS/CindyJS/blob/v0.8.2/src/js/libgeo/GeoRender.js#L66), else the demo file shows no lines at all.

File: src/js/libcs/OpDrawing.js [hunk 7]
```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
-
-        // Compute the roots of the x discriminant
-        // for points with horizontal tangents
-        sol = solveRealQuadratic(k00, -2 * k01, k02);
-        if (sol) {
-            for (i = 0; i < 2; ++i)
-                if (sol[i] >= 0 && sol[i] <= csh)
-                    boundary.push(
-                        mkp(-0.5 * (c11 * sol[i] + c10) / c20, sol[i]));
+    } while (--i > 0);
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js]: I consider the `do … while` harder to read than a corresponding `for`, so I'd suggest

```js
for (i = 0; i < n; ++i)
    if (points[i].t === "begin")
        previousBegin = points[i];
```

If forward iteration over these few points feels too wasteful, or you want to keep two levels of braces and the `next` variable for the sake of clarity, at least make it `for (i = n - 1; i >= 0; --i)`.

I do have to admit that I was worried for a moment about the fact that you excluded the `i === 0` case. Makes sense reading subsequent code, but feels like a bug at first. So although it would be fine to write the terminal condition as `i > 0` resp. the initialization as `i = 1` in the code above, I'd rather not do that to avoid the confusion.

File: src/js/libcs/OpDrawing.js [hunk 8]
```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
+    // For ovals compute the roots of the x and y discriminant for
+    // the horizontal and vertical tangent points respectively
+    var flats = k00 <= 0 ? [null, null] : [
+        solveRealQuadratic(k00, -2 * k10, k20),
+        solveRealQuadratic(k00, -2 * k01, k02)
+    ];
+
+    var points = [];
+    // Inside state of first corner at top left
+    var bInside = (sign(0, 0) < 0) === (det < 0);
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js]: How reliable is this choice here if the conic passes through the top left corner? I'd assume that we might end up with the wrong insideness flag for the whole boundary traversal in such a situation.

File: src/js/libcs/OpDrawing.js [hunk 9]
```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
...
+                break;
+            case "end":
+                if (df !== "D") csctx.lineTo(next.x, next.y);
+                else csctx.moveTo(next.x, next.y);
+                if (k00 < 0) drawArc(next, previousBegin);
+                break;
+            case "split":
+                drawArc(previous, next);
+                break;
         }
```
Comment (gagern) [inline src/js/libcs/OpDrawing.js:678]: Let me check whether I got this right. For hyperbolas you draw arcs from `end` to previous `begin`, but for ellpses you draw arcs from `end` to next `begin` via any `split` points in between, right? And if there are no boundary intersections, you jut connect `split` points. Furthermore you draw lines from `begin` via `corner` to `end` for filling and clipping operations. OK, makes sense.

I think I'd like to see a few more `closePath` calls. Particularly if someone should decide to draw conics with square line caps, having the paths closed should improve the rendering. Actually using square caps isn't as crazy as it might sound at first: if the current code is used to draw conics with really large line widths (the `size` parameter in the `63_Conic.html` example file), and the conic is nearly tangential to the boundary, then one can see the round line cap that's currently being used. Makes me wonder whether we should do all our intersections with a rectangle which is the viewport plus half the current line width, or something like that…

**Review by gagern — COMMENTED** [review]
Thanks for your input. There are a number of really nice changes. I'm somewhat concerned about the robustness of the decisions made by `doBoundary`, and would like your comment on that.

**Comment by gagern** [comment]
I've just changed added an `"F"` parameter to `drawgeoconic` to get an idea of how filling works. In the following example, which consists of two horizontal lines, I saw the whole canvas filled. The inside vs. outside distinction is undefined for degenerate conics, so I'd have expected either the area between the lines, or the area outside the lines to be filled.

    ["parallels", [-0.6666666666666666,1,-0.3333333333333333], [0.25,1,0.25], [0.3333333333333333,1,-0.3333333333333333], [1,0.5714285714285714,0.14285714285714285], [1,-0.42857142857142855,0.14285714285714285]],

File: src/js/libcs/OpDrawing.js (hunk 9 above)
Comment (gagern) [reply src/js/libcs/OpDrawing.js:678]: Sorry, but this does *not* make sense: for the `hyperbola2` example, this causes considerable off-screen drawing. That example is a single branch of a hyperbola which intersects the rectangle in four points. The correct arcs to draw would be 3 to 0 and 1 to 2. Getting this kind of combinatorics right was the point of the “sign at infinity” computation I did for the case of four points of intersection.

**Comment by gagern** [comment]
The `clipped wing` example also renders incorrectly. The number of `begin` vertices does not match the number of `end` vertices. Presumably because the conic passes through the bottom left corner, and numeric errors make both points of intersection fall outside the rectangle at that point.

I'm also no longer happy with your approach of always connecting `end` to `previousBegin` for hyperbolas, as it breaks the `hyperbola2` test case, even though the breakage isn't visually apparent. But see the comment in the review for details.

**Comment by gagern** [comment]
I'm thinking of how to decide whether to use the B-spline approach here or the non-rational refinement from my conicDrawing3 branch, I mean, the part before the `drawArc` call and the part after can be freely mixed between this branch here and the conicDrawing3 branch, right? So at some point, when we have settled for some version for the outer code, we will have to decide on one of these ways of drawing an arc, and I'm currently wondering how to make that decision.

Are there any special cases which we know that one algorithm can handle well but the other can not? Or should we assume that both algorithms work on all inputs (if they get called with the correct endpoints, which isn't always the case yet)? If they both work, there are different ways to compare them, and I'm not sure how to balance them against one another.

* **Code length:** currently I see [75 lines in conicDrawing3](https://github.com/gagern/CindyJS/blob/faea36bb24018c5fed77737525e9279f8847de16/src/js/libcs/OpDrawing.js#L705-L779) vs. [55 lines in your patch-2](https://github.com/elkins0/CindyJS/blob/3ab9bb0cd01d58f7e579a17831141e5471a60284/src/js/libcs/OpDrawing.js#L663-L718), so yours is shorter.
* **Code complexity:** I think the central idea of your code here is more complex than the non-rational approach, but after that there are fewer case distinctions and other weird control constructs. And in particular there is no more need to choose between different numeric roots of some quadratic equation.
* **Complexity of path:** I've just [created 1024 random](https://github.com/gagern/CindyJS/commits/conicTesting1) conics and let both algorithms draw them using a virtual canvas which just logs the draw commands. In this setup the conicDrawing3 code produced 28314 lines of output while this one here produced 30319 lines. But that's probably due to the incorrect handling of single hyperbola branches described above. If I take the `drawArk` and `refine` implementation here and plug them into the outer code from conicDrawing3 I see only 16626 lines of drawing commands. So it produces simpler paths.
* **Performance:** I've also timed the execution time. For the 1024 conics described above, your `drawArk` takes about 37ms while the one from conicDrawing takes 68. Which is approximately proportional to the ratio of path segments. So I'd conclude that computing a single Bézier arc takes about the same in both approaches, but drawing fewer makes your code more efficient.
* **Accuracy of path:** Of course it could well be that the reduced number of segments actually comes at a decreased quality of approximation. I'm wondering whether I should make the effort to take these random conics and analyze the resulting paths in a slow but reasonably accurate and reliable environment. Likely some numerical optimization or integration using mpmath. I'm not sure what metric to look for, though: do I want to find the maximal distance between a point on the drawn curve and its orthogonal projection to the actual conic? Or do I want the integral over these distances, or squared distances, or something like that? I can see reasons for each of these metrics.

File: src/js/libcs/OpDrawing.js (hunk 3 above)
Comment (elkins0) [reply src/js/libcs/OpDrawing.js]: Relying on the inside or outside state of the first corner is definitely a bad idea. I'm currently rethinking this.

File: src/js/libcs/OpDrawing.js (hunk 9 above)
Comment (elkins0) [reply src/js/libcs/OpDrawing.js:678]: The code now draws an arc from the end to previous begin only when hyperbola has its center within the clipping boundaries and there are two paths. In the rest of the situations the edge segments are connected by arc from their end to their beginning into one single path. Only while filling are these segments actually a part of the path.

**Comment by elkins0** [comment]
@gagern:
>But that's probably due to the incorrect handling of single hyperbola branches described above.

Yes, please redo using the commits since 10e5e5c8e795989d5941004f0bb3364ca603ef82
>Accuracy of path

 There is an alternative to consider. Since *Firefox* converts `quadraticCurveTo` into the equivalent  `bezierCurveTo` anyway, we could pass the curve thru the `G`  point instead of the `F` point by replacing `csctx.quadraticCurveTo(Bx, By, Cx, Cy);` with:
```JavaScript
var m = 4 / 3 * k;
var l = 1 - m;
csctx.bezierCurveTo(l * Ax + m * Bx, l * Ay + m * By, l * Cx + m * Bx, l * Cy + m * By, Cx, Cy);
```
Taking this concept a bit further, `quadraticCurveTo` could be incorporated as shown below. However `k <= 0.75` is not the correct measure. Something better is needed.
```JavaScript
    function refine(Ax, Ay, Bx, By, Cx, Cy, k, n) {
        if (k <= 0.75 || n > 5) {
            var m = 4 / 3 * k;
            var l = 1 - m;
            csctx.bezierCurveTo(l * Ax + m * Bx, l * Ay + m * By, l * Cx + m * Bx, l * Cy + m * By, Cx, Cy);
        } else {
            var Mx = 0.5 * (Ax + Cx);
            var My = 0.5 * (Ay + Cy);
            var kBx = k * Bx;
            var kBy = k * By;
            var j = 1 - k;
            var Gx = kBx + j * Mx;
            var Gy = kBy + j * My;
            var kr = 1 / (Math.sqrt(2 * j) + 1);
            ++n;
            refine(Ax, Ay, kBx + j * Ax, kBy + j * Ay, Gx, Gy, kr, n);
            refine(Gx, Gy, kBx + j * Cx, kBy + j * Cy, Cx, Cy, kr, n);
        }
    }
```
Here is a measurement experiment in *Mathematica* to look at the perpendicular distance from the desired curve to the qubic: 
```Mathematica
Manipulate[
 With[{P1 = pts[[1]], C = pts[[2]], P2 = pts[[3]], j = 1 - k}, 
  Block[{M, F, G, D, E, x1, y1, x2, y2, cx, cy, m = 4/3 k, C1, C2, Q1,
     dQ1, Q2, dQ2}, M = (P1 + P2)/2; F = (C + M)/2; G = k C + j M; 
   D = k C + j P1; E = k C + j P2; F = (C + M)/2; 
   G = k C + j M; {x1, y1} = P1; {x2, y2} = P2; {cx, cy} = C; 
   C1 = (1 - m) P1 + m*C; C2 = (1 - m) P2 + m*C; 
   Q1 = ((1 - u1)^3 P1 + 3 (1 - u1)^2 u1 C1 + 3 (1 - u1) u1^2 C2 + 
       u1^3 P2)/((1 - u1)^3 + 3 (1 - u1)^2 u1 + 3 (1 - u1) u1^2 + 
       u1^3);
   dQ1 = Cross[
     3 (C1 - P1 + 
        2 (-2 C1 + C2 + P1) u1 + (3 C1 - 3 C2 - P1 + P2) u1^2)]; 
   Q2 = ((u2 - 1)^2 (k - 1) P1 + 2 (u2 - 1) u2 k C + 
       u2^2 (k - 1) P2)/((u2 - 1)^2 (k - 1) + 2 (u2 - 1) u2 k + 
       u2^2 (k - 1)); 
   dQ2 = Cross[(2 (k - 
          1) (k P1 + (P1 - 3 k P1 + (k - 1) P2) u2 + (2 k - 1) (P1 - 
             P2) u2^2 + C k (2 u2 - 1)))/((u2 - 1)^2 (k - 1) + 
        2 (u2 - 1) u2 k + u2^2 (k - 1))^2]; 
   v = Solve[
     Eliminate[{Cross[Append[Q2, 1], Append[dQ2, 0]].{x, y, 1} == 
        0, {x, 
         y} == ((1 - u)^3 P1 + 3 (1 - u)^2 u C1 + 3 (1 - u) u^2 C2 + 
           u^3 P2)/((1 - u)^3 + 3 (1 - u)^2 u + 3 (1 - u) u^2 + 
           u^3)}, {x, y}], u]; 
   u1 = First@
     Select[u /. v, 
      Element[#, Reals] && 
        If[u2 < .5, 0 <= # <= 0.5, 0.5 <= # <= 1] &]; 
   distance = FullForm[EuclideanDistance[Q1, Q2]]; 
   area = 1/2 (-cy x1 + cy x2 + cx y1 - x2 y1 - cx y2 + x1 y2); 
   Graphics[{PointSize@0.01, Blue, Line[{P1, C, P2, P1}], Gray, 
     Line[{C, M}], Line[{D, E}], Line[{P1, E}], Line[{P1, G}], 
     Line[{G, P2}], Line[{D, P2}], Darker@Green, Point[G], 
     Text["G", G, {-2, -1}], 
     BSplineCurve[{P1, C, P2}, SplineWeights -> {1, k/(1 - k), 1}], 
     Dashed, Red, BezierCurve[{P1, C, P2}], Point[F], 
     Text["F", F, {-2, -1}], Green, 
     BSplineCurve[{P1, C1, C2, P2}, SplineWeights -> {1, 1, 1, 1}], 
     Red, Point[pts], Text["P1", P1, {0, -2}], Text["C", C, {-4, 0}], 
     Text["P2", P2, {0, -2}], Point[M], Text["M", M, {-2, -1}], 
     Point[D], Text["D", D, {2, -1}], Point[E], 
     Text["E", E, {-2, -1}], Magenta, Point[{Q1, Q2}], 
     InfiniteLine[Q2, dQ2]}, PlotRange -> {{0, 200}, {-100, 100}}, 
    PlotRangeClipping -> True, ImageSize -> {Full, Automatic}]]], {v, 
  0}, {u1, 0}, {{u2, 0.2}, 2^-53, 1. - 2^-53}, {{k, 0.75}, 0, 
  1. - 2^-53}, {distance, 0}, {area, 
  0}, {{pts, {{0, 0}, {100, 100}, {200, -100}}}, Locator}, 
 Bookmarks -> {"bottom" :> (pts = {{0, -100}, {100, 
        100}, {200, -100}}), 
   "bottom%10" :> (pts = {{90, -100}, {100, 100}, {110, -100}}), 
   "halfX" :> (pts = {{50, 0}, {100, 100}, {150, -100}}), 
   "quarterX" :> (pts = {{75, 0}, {100, 100}, {125, -100}})}]
```

**Comment by gagern** [comment]
I've finally managed to do a more thorough comparison of your code against mine. See [my conicTesting1 branch](https://github.com/gagern/CindyJS/commits/conicTesting1). The results can be summarized as follows:

1. Comparing your `drawArc` against mine, the precision is comparable. With the exception of three test cases (99, 305 and 535), the precision of both approaches differs by less than 0.001 px, measured in terms of the maximal distance between curve and conic. And even for these three cases, the maximal error is only 0.052.
2. There is no obvious trend for the differences of errors. The following figure shows the error (actually the smaller of the two errors) along the horizontal axis, and the difference in errors between vertically. The latter is on a much smaller scale, and appears to be just as likely to be positive than negative. Positive means your code is better, negative means mine is. Of the three outliers mentioned before, two are negative and one is positive.
  Difference in errors vs. minimal difference
3. The number of segments needed by the two `drawArc` implementations is almost always the same. There is only one test case, case 436, where my code uses 16 curved arcs but yours uses 13. The differences I observed before were likely due to bugs in the way I copied the implementation.
4. I also included your complete code as a third alternative in the tests. It has some serious outliers, where the maximal distance to the conic is *huge*. The most pronounced such case is this:

        ["Case 781", [-0.13482854887843132, -6.137520726770163, 1], [-8.664219973608851, -0.8530837446451187, 1], [4.688138976693153, 6.503501154482365, 1], [-5.400711786001921, 6.24084179662168, 1], [-7.9674973506480455, 6.623143093660474, 1]],

    As far as I can tell, all the affected cases draw a pair of lines where they should not. Can you please check whether you find a way to solve this?
5. The difference in performance appears to be gone now, too. So it likely was due to the same mistakes I made when copying the code.

On the whole I'd currently say that the outer code here needs fixing, and that the inner `drawArc` code from here and from #555 perform quite comparable, in terms of accuracy, path complexity and performance. So I think we really should make the decision based on code complexity.

Note that my test cases were generated by randomly choosing positions for the five points. It did not deliberately aim to produce corner cases. It might therefore well be that the two `drawArc` implementations differ significantly in such special situations.

**Comment by elkins0** [comment]
@gagern: Very nice work!
>4. ... As far as I can tell, all the affected cases draw a pair of lines where they should not. Can you please check whether you find a way to solve this?

Yes, sorry about that. It turns out that using the criteria for hyperbola reverse arc of requiring the conic center to be inbounds was too restrictive. Commit d6012897c9bc16b433ae6fa5433ce7ed3e17b694 widens the criteria by using the flats, only when they exist and both solutions for either x or y are outside and on either side of the boundaries, to restrict the use of the reverse arc.

Update:
Commit 08d99b8db412c18f23ca5ff57c3c82674b87aafd tweaks an inadvertent change that ended up splitting parabolas unintentionally.

Commit 1b4a30e547dd7bb5557ec9c06a1587ab6bfd6401 has all the changes needed to update the [conicTesting1](https://github.com/gagern/CindyJS/commits/conicTesting1) branch.

**Comment by gagern** [comment]
I re-ran my checks, and it seems your code is working on all examples. I've merged it into my conicDrawing4 branch now, to simplify the situation somewhat. Now we only have to decide between conicDrawing3, conicDrawing4 and a mixture with the outer code from conicDrawing3 and the drawArc implementation from conicDrawing4. I haven't really looked in detail at your latest changes to the outer code, and I wanted to adapt some of your ideas to conicDrawing3 first. I guess I'll discuss the state of affairs with the rest of the ream next Monday, and hope to reach a decision soon after that.
